### ISSUE TITLE
[1.4.x] bug w/ FastRender, inject-data after upgrade to 1.4 and webapp >1.3.10

### ISSUE THREAD

**Issue by rigconfig** [meteor/meteor#7992]
# Summary

I'm using [reactrouter:react-router-ssr](https://github.com/thereactivestack/meteor-react-router-ssr) which exposes a data de/rehydration api via ReactRouter server-side-rendering. It uses [FastRender](https://github.com/kadirahq/fast-render/) to accomplish this.

The [data dehydration callback](https://github.com/rigconfig/meteor-webapp-bug/blob/master/imports/Router.jsx#L18) works in Meteor 1.3.5.1, but when I upgrade to Meteor >=1.4 FastRender no longer injects data into the script tag on the client; therefore the rehydration callback doesn't receive the server-side data it was expecting.

Others are reporting a similar problem after updating to 1.4: [here](https://github.com/kadirahq/fast-render/issues/186) and [here](https://github.com/thereactivestack/meteor-react-router-ssr/issues/102).
# Clues

My best guess yet is that it's `webapp` > 1.3.10

**Reason:** When I manually updated Meteor from 1.3.5.1 to 1.4 _(aka 1.4.0)_, FastRender still worked as long as I didn't force package updates with `meteor update --packages-only`, so I was still using webapp@1.3.10.

When I updated the packages for Meteor 1.4, `webapp` updated to 1.3.12 and then FastRender stopped injecting it's data on the client.

I've also tried cloning reactrouter:react-router-ssr and fast-render as local packages and bumping their `webapp` and `connect` versions to no avail. But I'm asuming it's webapp simply because [fast-render](https://github.com/kadirahq/fast-render), [picker](https://github.com/meteorhacks/picker), [inject-data](https://github.com/meteorhacks/inject-data/) all use or depend on `webapp`, `connect` and `http` packages to modify the server's middleware response for server-side-rendering.
# Reproduction Recipe
## Produce success condition w/ Meteor@1.3.5.1
1. `git clone https://github.com/rigconfig/meteor-webapp-bug.git`
2. `cd meteor-webapp-bug/`
3. `npm i && meteor run` _(do not update packages, just run)_
4. Open: [http://localhost:3000/](http://localhost:3000/)
5. Open Chrome DevTools console _(or other inspector)_
6. You should see one client console.log which reads:
   - `rehydrateHook: Object {testprop: "this is a test"}`
7. Also, a single server console.log which reads:
   -  `dehydrateHook: { testprop: 'this is a test' }`
8. Check `/.meteor/versions` and webapp version should be `webapp@1.2.11`
9. This is an example of successful operation and expected behaviour.
## Update to Meteor@1.4 (still successful)
1. `meteor update --release 1.4`
2. `meteor run` _(again, do not update packages or meteor)_
3. The same success condition w/ console logs should be observed.
4. Check `/.meteor/versions` and webapp version should now be `webapp@1.3.10`

So now we've upgraded meteor to 1.4 and fast-render _(and the router's hydration funcs)_ are still working on both client and server as expected.
## Introduce breaking change
1. `meteor update --packages-only`
2. `meteor run`
3. Refresh browser of app and in client console you should see failure:
   - `rehydrateHook: undefined`
4. Also a single server console.log which still correctly reads:
   -  `dehydrateHook: { testprop: 'this is a test' }`
5. Check the browser source of the html page on client and there should no longer be any "fast-render" present in the script injection tag. e.g. `<script type="text/inject-data"> (...) fast-render-data (...)`
6. Check `/.meteor/versions` and webapp version should be `webapp@1.3.12` _(from 1.3.10 previously)_

If you need anything else let me know. In the meantime, I'll continue troubleshooting and update if I find anything else. Thanks for your help.
# Atmosphere Packages

``` console
meteor-base             # Packages every Meteor app needs to have
mobile-experience       # Packages for a great mobile UX
mongo                   # The database Meteor supports right now
blaze-html-templates    # Compile .html files into Meteor Blaze views
reactive-var            # Reactive variable for tracker
jquery                  # Helpful client-side library
tracker                 # Meteor's client-side reactive programming library

standard-minifier-css   # CSS minifier run for production mode
standard-minifier-js    # JS minifier run for production mode
es5-shim                # ECMAScript 5 compatibility for older browsers.
ecmascript              # Enable ECMAScript2015+ syntax in app code
reactrouter:react-router-ssr   # Implies: react-meteor-data
```
# NPM Packages

``` json
  "dependencies": {
    "meteor-node-stubs": "~0.2.0",
    "react": "^15.3.0",
    "react-addons-pure-render-mixin": "^15.3.2",
    "react-dom": "^15.3.0",
    "react-mixin": "^3.0.5",
    "react-router": "^2.0.1"
  }
```
### When reporting a bug, please be sure to include the following:
- [x] A descriptive title
- [x] What version of Meteor you're using, and the platform(s) you're running it on
- [x] What packages or other dependencies you're using
- [x] The behavior you expect to see, and the actual behavior
- [x] An _isolated_ way to reproduce the behavior (example: GitHub repository with code isolated to the issue that anyone can clone to observe the problem)

**Comment by CharlesOy** [comment]
I don't know if it's for the same reason, but flow-router-ssr doesn't work either.
I mean it's working, but without ssr and fast-render.

**Comment by rigconfig** [comment]
**Update:** Getting warmer...

Looks like the immediate problem may be in `inject-data` package. If I comment out these lines it works in 1.4 and fast-render injects data again as expected:


```javascript
/* File: /packages/inject-data/lib/server.js */

InjectData._hijackWriteIfNeeded = function(res) {  
  // if(res._writeHijacked) {
  //   return;
  // }
```
Source: https://github.com/meteorhacks/inject-data/blob/master/lib/server.js#L25

I'm investigating further to see what changed and understantd why that's there and what it's doing? It's just a monkeypatch at this point.

**Comment by rigconfig** [comment]
**Update:** So, from the code above it looks like `res._writeHijacked ` is set to `true` the first time `InjectData._hijackWriteIfNeeded()` is called. Then all subsequent calls to that function meet the truthy condition in that code I commented out, which causes it to simply be returned silently w/out executing.

@simsim0709 has a different take [here](https://github.com/kadirahq/fast-render/issues/186#issuecomment-257197096).

I'm not familiar with what's going on further up the stack, so if anyone else wants to chime in feel free.

For now, I've [forked](https://github.com/rigconfig/inject-data) the inject-data repo w/ the fix:
```console
cd packages/
git clone https://github.com/rigconfig/inject-data
meteor run
```
## Level Complete
[![Level Complete](https://67.media.tumblr.com/5aa6ad4c78744f74e121dc6c5ec8574c/tumblr_n3q0x1oHAh1rj5xamo1_500.gif)](https://youtu.be/pU-DfRSFWbA?t=2m54s)

**Comment by CharlesOy** [comment]
Thank you very much.
It seems `res._writeHijacked` ensured that `InjectData._hijackWriteIfNeeded()` could only be invoked one time, is it because `InjectData._hijackWriteIfNeeded()` has been invoked somewhere before real injected data comes in?

**Comment by comus** [comment]
the main problem is picker and webapp

first, in `fast-render/lib/server/routes.js`

```js
var fastRenderRoutes = Picker.filter(function(req, res) {  // <-- fast render use picker
  return IsAppUrl(req);
});
fastRenderRoutes.middleware(Npm.require('connect').cookieParser());
fastRenderRoutes.middleware(function(req, res, next) {
  FastRender.handleOnAllRoutes(req, res, next); // <-- InjectData.pushData in this function, and pushData will call InjectData._hijackWriteIfNeeded to hack res.write
});
```

and then, in `picker/lib/instance.js` show that

```js
WebApp.rawConnectHandlers.use(function(req, res, next) { // <-- it use webapp 's rawConnectHandlers
  Picker._dispatch(req, res, next);
});
```

so, what happen in webapp update, just a little changes
you can see more here: https://github.com/meteor/meteor/pull/7753

the webapp update move `connect.compress()` after `rawConnectHandlers`

```
app.use(rawConnectHandlers); // <-- InjectData._hijackWriteIfNeeded hack res.write here
app.use(connect.compress()); // <-- connect.compress()
```

finally, the program will run `connect.compress()` first,
and `InjectData._hijackWriteIfNeeded` 's res.write function cannot read the compressed html,
so InjectData cannot find the first `<script` position in html to add its fast render script

**Comment by abernix** [comment]
Thanks for reporting this problem originally.  While the change that was PR'd into `webapp` has caused this problem (for users of `kadira:fast-render`) the problem itself lies in that third-party package.

I'd venture to guess that it's unlikely that the PR will be reverted as it fixes an important issue with `rawConnectHandlers` in that they were not truly "raw" without it.  The new method provides a more clean ways to hook into core functionality for third party packages.

If a third-party package was monkey-patching Meteor internals, then the third-party package will need to be fixed.  There are new `dynamicHead` and `dynamicBody` facilities in place (#3860) that seem like they might be helpful in doing this.

I see in the other issues about this problem (e.g. kadirahq/fast-render#186) that its being claimed that the package is abandonware, but I don't see anyone having submitted a PR yet.  Kadira may not be doing active development on it but they have merged well-written PRs in recent times on other projects of theirs and I think the chances of a PR getting accepted for this would be increased if it was carefully implemented to not be a breaking change for pre-1.4 users.  I'm sure if someone took the opportunity to do that, it would be appreciated by the community.

Closing, as this doesn't appear to be a bug in Meteor but rather a monkey-patched implementation which has broken.  Happy to reopen if someone can demonstrate that this is a bug in Meteor!

**Comment by abecks** [comment]
I have [submitted a PR](https://github.com/meteorhacks/inject-data/pull/11) to meteorhacks:inject-data that adds support for Meteor 1.4.2+. It is backwards compatible with older versions of Meteor aswell.
